framework/helpers/BaseJson.php

```diff
@@ -179,16 +192,28 @@ protected static function processData($data, &$expressions, $expPrefix)
                 $data = $data->toArray();
             } elseif ($data instanceof \SimpleXMLElement) {
                 $data = (array) $data;
+
+                // Avoid empty elements to be returned as array.
+                // Not breaking BC because empty array was always cast to stdClass before.
+                $revertToObject = true;
             } else {
+                /*
+                 * $data type is changed to array here and its elements will be processed further
+                 * We must cast $data back to object later to keep intended dictionary type in JSON.
+                 * Revert is only done when keepObjectType flag is provided to avoid breaking BC
+                 */
+                $revertToObject = self::$keepObjectType;
+
                 $result = [];
                 foreach ($data as $name => $value) {
                     $result[$name] = $value;
                 }
                 $data = $result;
-            }
 
-            if ($data === []) {
-                return new \stdClass();
+                // Avoid empty objects to be returned as array (would break BC without keepObjectType flag)
+                if ($data === []) {
```

Comment:
it's unuseful - check `if (is_object($data))`

Reply:
$data is array there and never object. This check was there to ensure that empty objects were not encoded as empty array (which was only half of the issue - rest of it is fixed now)

Reply:
@zebraf1 Note: you can't keep object type if passed something like this `new ArrayObject([1, 2], 1)` because this object stored index array. Old version keep type only for empty objects.

Reply:
@zebraf1 i try say: in case where method convert object to array, array must be revert to object, flag no need:
~~~php
if (is object($data)) {
   // ...
   if (!is_object($data)) {
       $data = (object) $data;
   }
}
~~~

Reply:
Ah yes, this is what I wanted to do at first, because it is a bug. But owners said it would be breaking change, so I had to put it under a flag. Then developers can choose to include this fix or not - it must not break existing applications who upgrade Yii version.